File: backend/app/schemas/certificate.py
"""
Certificate Schemas

Pydantic schemas for certificate management API.
"""
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class CertificateRegisterRequest(BaseModel):
    """Request to register a new certificate."""
    type: str = Field(..., description="Certificate type (e.g., PKI_OVERHEID)")
    storage_ref: str = Field(..., description="Path or reference to certificate file")
    passphrase_ref: Optional[str] = Field(None, description="Reference to passphrase (env var or secret)")
    friendly_name: Optional[str] = Field(None, description="User-friendly name for the certificate")
    purpose: Optional[str] = Field(None, description="Purpose of certificate (e.g., BTW_SUBMISSION)")
    
    @field_validator('type')
    @classmethod
    def validate_type(cls, v):
        allowed_types = ['PKI_OVERHEID']
        if v not in allowed_types:
            raise ValueError(f"Certificate type must be one of {allowed_types}")
        return v
    
    @field_validator('storage_ref')
    @classmethod
    def validate_storage_ref(cls, v):
        if not v or len(v.strip()) == 0:
            raise ValueError("storage_ref cannot be empty")
        return v


class CertificateMetadata(BaseModel):
    """Certificate metadata extracted from the certificate file."""
    fingerprint: str
    subject: str
    issuer: str
    serial_number: str
    valid_from: datetime
    valid_to: datetime


class CertificateResponse(BaseModel):
    """Response containing certificate information."""
    id: UUID
    administration_id: UUID
    type: str
    storage_ref: str
    has_passphrase: bool  # Whether passphrase_ref is set (don't expose the actual ref)
    fingerprint: str
    subject: str
    issuer: str
    serial_number: str
    valid_from: datetime
    valid_to: datetime
    friendly_name: Optional[str]
    purpose: Optional[str]
    created_at: datetime
    created_by: UUID
    updated_at: datetime
    is_active: bool
    is_valid: bool  # Computed: whether cert is currently valid
    days_until_expiry: int  # Computed: days until expiry
    
    class Config:
        from_attributes = True


class CertificateListResponse(BaseModel):
    """Response containing list of certificates."""
    certificates: list[CertificateResponse]
    total: int


class CertificateRegisterResponse(BaseModel):
    """Response after registering a certificate."""
    certificate: CertificateResponse
    message: str = "Certificate registered successfully"


class CertificateDeleteResponse(BaseModel):
    """Response after deleting a certificate."""
    message: str = "Certificate deleted successfully"


class SigningRequest(BaseModel):
    """Request to sign XML content."""
    xml_content: str = Field(..., description="XML content to sign")
    certificate_id: UUID = Field(..., description="ID of certificate to use for signing")


class SigningResponse(BaseModel):
    """Response containing signed XML."""
    signed_xml: str = Field(..., description="Signed XML content")
    signature_info: dict = Field(..., description="Information about the signature")
    certificate_fingerprint: str = Field(..., description="Fingerprint of certificate used")

File: src/lib/api/certificate.ts
// ============ Certificate Management API ============
// Extracted from src/lib/api.ts as part of the api.ts decomposition.
// Behavior, endpoints, and response shapes are unchanged.

import { api } from '../api'

export interface CertificateResponse {
  id: string
  administration_id: string
  type: string
  storage_ref: string
  has_passphrase: boolean
  fingerprint: string
  subject: string
  issuer: string
  serial_number: string
  valid_from: string
  valid_to: string
  friendly_name: string | null
  purpose: string | null
  created_at: string
  created_by: string
  updated_at: string
  is_active: boolean
  is_valid: boolean
  days_until_expiry: number
}

export interface CertificateListResponse {
  certificates: CertificateResponse[]
  total: number
}

export interface CertificateRegisterRequest {
  type: string
  storage_ref: string
  passphrase_ref?: string
  friendly_name?: string
  purpose?: string
}

export interface CertificateRegisterResponse {
  certificate: CertificateResponse
  message: string
}

export interface CertificateDeleteResponse {
  message: string
}

export const certificateApi = {
  list: async (
    clientId: string,
    includeExpired?: boolean
  ): Promise<CertificateListResponse> => {
    const params: Record<string, unknown> = {}
    if (includeExpired) params.include_expired = includeExpired
    const response = await api.get<CertificateListResponse>(
      `/accountant/clients/${clientId}/certificates`,
      { params }
    )
    return response.data
  },

  get: async (clientId: string, certificateId: string): Promise<CertificateResponse> => {
    const response = await api.get<CertificateResponse>(
      `/accountant/clients/${clientId}/certificates/${certificateId}`
    )
    return response.data
  },

  register: async (
    clientId: string,
    request: CertificateRegisterRequest
  ): Promise<CertificateRegisterResponse> => {
    const response = await api.post<CertificateRegisterResponse>(
      `/accountant/clients/${clientId}/certificates/register`,
      request
    )
    return response.data
  },

  delete: async (clientId: string, certificateId: string): Promise<CertificateDeleteResponse> => {
    const response = await api.delete<CertificateDeleteResponse>(
      `/accountant/clients/${clientId}/certificates/${certificateId}`
    )
    return response.data
  },
}
